File: kotlin/friends-tree/src/main/kotlin/com/pret/friend/tree/repository/UserFriendsRepository.kt
package com.pret.friend.tree.repository

import com.pret.friend.tree.data.DataComponent
import com.pret.friend.tree.abstractions.User
import org.springframework.stereotype.Repository

@Repository
class UserFriendsRepository(
    private val dataComponent: DataComponent
) {

    fun getFriendsById(userId: Int) =
        dataComponent.friendships().filter { it.first == userId }
            .map { friendShipPair ->
                getUserById(friendShipPair.second)
            }

    fun getUserById(userId: Int): User = dataComponent.users().first { it.id == userId }
}

File: nodejs/src/repository/userFriendsRepository.ts
import { User } from "../types/general";
import { users } from "../../external-friends-service/users";
import { friendships } from "../../external-friends-service/friendships";

const userFriendsRepository = {
  getUserById(id: string): User | undefined {
    return users.find((u) => u.id === id);
  },

  getFriendsById(userId: string): User[] {
    return friendships
      .filter(([a]) => a === userId)
      .map(([_, friendId]) => this.getUserById(friendId as string))
      .filter((u): u is User => u !== undefined);
  },
};

export default userFriendsRepository;
